src/main/java/com/hierynomus/smbj/auth/NtlmAuthenticator.java

```diff
@@ -60,19 +61,24 @@ public NtlmAuthenticator create() {
         }
     }
 
-    public long authenticate(Connection connection, AuthenticationContext context) throws TransportException {
+    public Session authenticate(Connection connection, AuthenticationContext context) throws TransportException {
```

Comment:
Here's an issue.  In the NtlmAuthenticator, this can make sense because there is only a simple exchange.  However, in a more general GSSAPI authentication scenario where you have multiple rounds, it seems like the suggestion from the MS-SMB2 document is that we should have a separate session table for sessions that are in "authentication state", but that they are full session objects.  I don't fully understand why that is, but I haven't looked carefully at the situation.

So, while I understand the idea of separating SessionDetails from Session, I'm not convinced that this is the best separation.  So I'd like to punt this decision until we understand this better.

My gut feeling is that this is all driven from a (possibly unavoidable) messy protocol design, where we have a "stack" of protocol contexts (eg, connection, session, treeconnect, share, etc.) in which all actions will take place.  So that means that all of those contexts will need to be split into "details" and "code/state" objects, if we really care about the architectural separation.  Otherwise, we just accept the cycle, on the basis that all elements of the stack are actually required by the protocol (and by Java limitations).